### ISSUE TITLE
Add syntax for a fast-fail/strict expansion to string interpolation

### ISSUE THREAD

**Issue by jimbobmcgee** [PowerShell/PowerShell#24218]
### Summary of the new feature / enhancement

As a user, I frequently use environment variables to refer to (for example) certain filesystem locations.  Many times, the commands I use are pasted into the command-line from a file of snippets, rather than a full-blown script.  This can lead to problems across multiple systems, if such a variable has not been set.  

For example, if I were to do `. "${env:TOOLSROOT}\this-tool.ps1"` with *TOOLSROOT* having not been set, I would attempt to source a file from the root of the current directory that, *at best*, does not exist and, *at worst*, exists and is not the file I intended.  

Now, in full-blown scripts, this is easy enough to circumvent with a basic `if ([String]::IsNullOrEmpty("${env:TOOLSROOT}"))` or even a `Set-StrictMode` for normal variables (but&mdash;I note&mdash;not for environment variables), but is a bit verbose for a snippet or hand-rolled command-line.

POSIX shells, such as Bash, have a number of punctuation/operators in their interpolation syntax which can alter the resulting value, depending on whether the interpolated variable is unset or empty, such as `${AAA-BBB}` (emit value *BBB* if variable *AAA* is unset, else emit the value of variable *AAA*) or `${AAA:-BBB}` (emit value *BBB* if variable *AAA* is empty or unset, else emit the value of variable *AAA*).  I am not proposing to implement all of these, but there may be some use in implementing some aspect of the `?` operator.

Technically, in POSIX, `${AAA?:BBB}` means *exit with an error string 'BBB' if the variable AAA is unset, else emit the value of variable AAA*.  I am thinking we could apply this `?` symbol to PowerShell's variable interpolation, and have it fail the expansion with an error, if the variable is not set or results in an empty-string expansion.  This would not be affected by `Set-StrictMode`, and would always error if empty/unset and always emit the value if successful.

### Proposed technical implementation details (optional)

Allow for `"${?scope:varname}"` to throw an error if `${scope:varname}` is unset or would expand to the empty string.  Alternative syntax might be "${scope?:varname}", if that is easier for the existing parser to process.

e.g.
```
$env:a = "abc123"
Write-Host "${env:a}"
>> abc123
Write-Host "${?env:a}"
>> abc123
Write-Host "${?env:b}"
>> The variable '$env:b' cannot be retrieved because it has not been set.
At line:1 char:13
+ Write-Host "${?env:b}"
+             ~~
    + CategoryInfo          : InvalidOperation: (b:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined

$x = "xyz321"
Write-Host "$x"
>> xyz321
Write-Host "${local:x}"
>> xyz321
Write-Host "${?local:x}"
>> xyz321
Write-Host "${?local:y}"
>> The variable '$local:y' cannot be retrieved because it has not been set.
At line:1 char:13
+ Write-Host "${?local:y}"
+             ~~
    + CategoryInfo          : InvalidOperation: (b:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined
```

I have picked the `?` punctuation because it matches the POSIX operator (although that isn't a requirement), and because I don't *think* it is already allowed in existing namespaces, so its introduction should not affect prior work.

Note that I am not expecting to be able to customise the error message thrown as a result of a failed expansion (as one can with the POSIX equivalent), only that it should raise an error and prevent execution with an understandable message.  Whether that error is *terminating* or not is for discussion.

**Comment by kasini3000** [comment]
Forcibly shortening the code length will lead to increased complexity.
source  /path/to/your-environment-variable.ps1
[System.Console]::Title = 'your environment variable load'
or
play music
or 
read string by speaker

**Comment by jhoneill** [comment]
`. "$env:TOOLSROOT\this-tool.ps1" `   (it really doesn't need the braces) will run from the root of the current drive, it won't run 
`.   .\this-tool.ps1 `  (this is frequently a problem with using $PSScriptRoot and then using [F8] to step through the code in an idea instead of running the script from a a file - I have taken to using "If ($PsscriptRoot) {Push-location $psscriptRoot}" (with a pop a then of the script and run use $pwd in place of $psScriptRoot)
We get an error and go "DOH" and put it right.   now let's say we have a new operator which returns the value or throws 
`.  "$($env:ToolsRoot ?!)\this-tool.ps1 ` 
And let's assume we've got over the hurdle that scripts which don't work in older powershell are held to be a bad thing.

Will we remember to to type this to give a different error "your variable is null" instead of "couldn't find the file in the root" ?
Because I don't think anyone will think "I had better use the longer syntax in case I didn't set the variable AND the path points to somewhere I don't want" 
 
Just as we have `@" "@ `  for multi-line interpolated strings we could have something else for fast-failing strings, but again will we remember - decades of muscle memory have me typing `"` when `'` would do. 
And is the swap of error messages a sufficient benefit?  Given the only time a script runs is when the unintended version is in the root of the file system, the person who puts an old version in the wrong place and forgets to set the variable before use is probably no  going to remember to use the defensive syntax.

**Comment by jimbobmcgee** [comment]
@kasini3000
> Forcibly shortening the code length will lead to increased complexity.

Adding any new feature increases complexity.  More-verbose code is not always less complex, either.  Shorter, consise code which uses well-documented features tend to increase usability.

**Comment by rhubarb-geek-nz** [comment]
> Shorter, concise code which uses well-documented features tend to increase usability.

I suggest not always. You can document the PowerShell aliases as much as you like but they still come across as hieroglyphic gibberish in comparison to the full verb-noun functions and cmdlets. I suggest the full verb-noun form, while more verbose, makes the code more readable and maintainable.

**Comment by jimbobmcgee** [comment]
> > Shorter, concise code which uses well-documented features tend to increase usability.
> 
> I suggest not always. You can document the PowerShell aliases as much as you like but they still come across as hieroglyphic gibberish in comparison to the full verb-noun functions and cmdlets. 

And yet, l'd bet that most people do not type `Get-ChildItem` into the interactive shell that often. 😀

However, we are getting off-topic.  This wasn't supposed to be a treatise on the pros and cons of `Noun-Verb`, nor less-verbose code *in general*.

I am very specifically interested in a string interpolation syntax which *errors* if a variable is empty or unset. `Set-StrictMode` doesn't cover all cases (environment variables, in particular, are not subject to strict checking), and the expansion of `"$(if ([String]::IsNullOrEmpty("${env:FolderPath}")) { throw 'expand fail' } else { "${env:FolderPath}" })\file.ext"` is utterly tortuous (if it even works) and could obviously be made more user-friendly.

**Comment by rhubarb-geek-nz** [comment]
>  and could obviously be made more user-friendly _by syntax_ without loss of readability or intent.

Unlike compiled languages, scripting languages have to be very conservative in changes. For instance your proposal would mean any scripts using this feature would need a "#Required -Version 7.5" to prevent execution where the feature has not been implemented.

**Comment by MartinGC94** [comment]
> For instance your proposal would mean any scripts using this feature would need a "#Requires -Version 7.5" to prevent execution where the feature has not been implemented.

So what? The same can be said for any new feature. By that logic the team might as well stop all development on PowerShell because any new feature or bug fix they make could cause scripts to behave differently in older versions VS newer versions.  
PowerShell has already added new "convenience" syntax features that are incompatible with older versions like ternaries and pipes at the start of new lines. Neither feature were strictly needed, they were just added because they were nice to have.

I personally see a lot of value in OPs suggestion and I think it would be nice if something like this was added. I'm just not sure how best to do this. Strict mode seems like the most straight forward way but because of the unfortunate decision to use versions rather than individual feature toggles, I think the usage would be quite limited.

**Comment by jhoneill** [comment]
> > For instance your proposal would mean any scripts using this feature would need a "#Requires -Version 7.5" to prevent execution where the feature has not been implemented.
> 
> So what? The same can be said for any new feature. 
This is true and there is some resistance to creating code which might look syntactically correct but fails in down level versions. 

> I personally see a lot of value in OPs suggestion 
As described I think people will use the "normal" syntax in the use case set out. 

There could be either
* A new kind of quotes for 'interpolate but fail fast' e.g. `?" xxx  ?"`  
* A new "null operator" to condense `$x ?? $(throw "Bad null")`  it's probably easier to add (e.g.) a ?! operator than to add things to string interpolation and it's usable in more places, but more verbose in use. 

However this has the ring of "I typed something wrong, and I thought wouldn't it be better if that could be handled by the some new language syntax?"   And the answer to the that is almost always "No". 
* The mistake causes an error in nearly all cases (it only rarely runs something that shouldn't run - see my comment above) and the new syntax to change the error message isn't a good way to use developers' time. 
* Muscle memory, the way people are taught, and above all the failure to anticipate that the variable might be null all mean the new syntax will only rarely be used, and in a script we already check for null and produce an error suitable for the context ("Path needs to be valid" not "Bad null at line 42").   

compare these two bits of code for handling bad directory and a failure to open the file. 
```
if (-not $dir) {throw 'Need a directory} 
$foo = get-content "$dir\foo.txt 
if (-not $foo) {Throw 'Error opening file'} 
 ```
and 
```
try {$foo = get-content "$($dir ??)\foo.txt" }
catch {
    if ($_  is_bad_null ) {throw "need a directory"}
    else {throw "Error opening file"}
}
```

I'd say the first is easier to understand, and of course wouldn't be confined to a specific version 
   
An error_if_null operator might still have value... but I'm not seeing it for the examples given

**Comment by jimbobmcgee** [comment]
@rhubarb-geek-nz:
>For instance your proposal would mean any scripts using this feature would need a "#Requires -Version 7.5" to prevent execution where the feature has not been implemented.

Not really.  If a script contains a new feature or syntax that is not recognised by an older version of the shell, then it fails to parse.  That's entirely normal.  It is only necessary to gate/guard new features which might create breaking changes to reasonable existing code.

Consider the proposed syntax `${?scope:variable}`, such as `${?env:test}` or `${?local:test}`: Noone is going to use that inside of an existing/old script, because it doesn't do anything.  You can't *set* a variable called `${?local:test}` in any existing shell version (you get a 'DriveNotFound' error); if you attempt to *get* that variable, you get `$null`; and if you *interpolate* it, you get the empty string.

Now, consider this feature has been implemented, as described, in a *new* shell version:  If you *interpolate* `"${?local:test}"`, you will now either get the stringified value of `${local:test}` or an error.  

I'm not seeing a danger here.

*If* I were saying that I wanted the existing behaviour of `${local:test}` or `${env:test}` to be changed to emit an error if it were interpolated, then your argument against change would be understandable.  However, the behaviour of `${?local:test}` is currently *undefined*, noone can expect to rely on the current `$null`-returning behaviour.

Defining a behaviour for something that did not have a definition before is *exactly* what a new feature *is*.

**Comment by jhoneill** [comment]
> Consider the proposed syntax `${?scope:variable}`, such as `${?env:test}` or `${?local:test}`: No-one is likely to be using that

@jimbobmcgee   Do you think this syntax will ever be used ? 

The `${}` Syntax is almost never used.  What actually happens at the command line is people type 
`gc  $foo\bar.txt`  
and get 
`Cannot find path 'C:\bar.txt ' because it does not exist.` 

They simply won't convert to typing `gc  {?local:foo}\bar.txt`  on the off chance that they have forgotten to set $foo.  If that possibility occurs to them they'll probably used do `$foo` to check its value first.  

In a function it's really only applicable to external variables - local variables are set in the function, parameters can be made mandatory -  but a "variable is null" error is not giving a better result , assuming that people think to code it, instead of thinking "oh that environment variable might not be set, what should I do in that case"

In short, if you realize you might need it,  you've probably already thought of an other method to solve the problem without it, so you don't need it.  Catch-22 for a new feature, no ?

**Comment by jimbobmcgee** [comment]
@jhoneill 
> Do you think this syntax will ever be used ?

If I didn't think it would be used, I wouldn't have suggested it.  There are many things that I expect very few people *use*, but they *exist*, and that doesn't mean that they *shouldn't* &ndash; find me a command or cmdlet that uses the approved verb *Ping*, for example (`Get-Command -All -Verb Ping` returns nothing in *my* WinPS5.1), but *Ping* is still a valid response from `Get-Verb`.


> The `${}` Syntax is almost never used

You seem fixated on the braces.  It's not about the braces.  The braces are not as important as the functionality.  For what it is worth *I* use the braces, *especially* in string interpolation, and I would recommend anyone else do the same.  The braces exist for a reason and they should be used for exactly the reason they were invented.  

If you think a form of the string-interpolation syntax I am requesting should also work without the braces, then *I am all for that*, although it *might* be easier on the parser logic if we require the braces when using the proposed syntax, *as written* &ndash; `"$?scope:var"` would be a breaking change for anyone who thought they were interpolating variable `"$?"`, for example, while `"${?scope:var}"` *shouldn't* fall foul of that.

I'm also open to *any other syntax* that passes the *can be pasted from a text file full of one-liners* requirement, which is the reason for the suggestion.  And *no*, running a script or sourcing a profile that exposes a function to do this *does not pass that requirement*.


> In short, if you realize you might need it, you've probably already thought of an other method to solve the problem without it

Of course I have, it's just that the other methods *suck* and, since this is an open forum with a dedicated *Issue-Enhancement* tag, I *assumed* that suggestions were&mdash;you know&mdash;*welcome*. 😄

**Comment by jhoneill** [comment]
> @jhoneill
> 
> > Do you think this syntax will ever be used ?
> 
> If I didn't think it would be used, I wouldn't have suggested it. 

OK... I'll go back to what I said before, people often come up with suggestions, along the lines of "I had a problem and therefore the language should change". Those suggestions shouldn't be _unwelcome_ but often the result is, yes that would solve your problem, but it won't make a lot of other people wonder how they managed without it. There are more possible features than the team can implement, so would they please more people by implementing this one or another one?  And sometimes the answer is "what you're talking about is part of a bigger problem which could do with a solution"

I see a catch 22. I inherited scripts with `.  $PsScriptRoot\library.ps1 `  and when I use [F8] to run it to looks for the library in the root. This would change the error I get. If I have the presence of mind to handle the absence of the variable, I probably do something other than causing a different error.  


>There are many things that I expect very few people _use_, but they _exist_, and that doesn't mean that they _shouldn't_ – find me a command or cmdlet that uses the approved verb _Ping_, for example
My last client had implemented Ping-_serviceName_ for the APIs they provided to internal apps. 



> > The `${}` Syntax is almost never used
> 
> You seem fixated on the braces. It's not about the braces. The braces are not as important as the functionality. For what it is worth _I_ use the braces, _especially_ in string interpolation, and I would recommend anyone else do the same. The braces exist for a reason and they should be used for exactly the reason they were invented.

The purpose of the braces was to allow names which would otherwise be illegal. e.g.
```
> New-Variable -Name "@@" -value "boo"
> $@@                                                                                                                                        
$@@: The term '$@@' is not recognized as a name of a cmdlet, function, script file, or executable program.> ${@@}
boo
```
It's one way to solve a problem of string interpolating where variable names run into text e.g.
` "$gw`gigawatts" ` is ok, but `"$co2`tons" uses a tab. Most cases use concatenation or a -f to dodge this. 

 
> If you think a form of the string-interpolation syntax I am requesting should also work without the braces,

I'm wondering if it needs more than a generic solution - just as we have  `$private:foo.prop`  where  private is a label which looks like a scope, we could have `$strict:foo.prop`

Or we could have an operator something like ??! which throws if any of the objects passed to it is null  - and I can see uses for `$a, $b, $c ??!`  or even `$array ??!` 

Or we could have some special syntax for a fast failing interpolated string like @@"xxx"@@ 
 
> > In short, if you realize you might need it, you've probably already thought of an other method to solve the problem without it
> 
> Of course I have, it's just that the other methods _suck_ and, since this is an open forum with a dedicated _Issue-Enhancement_ tag, I _assumed_ that suggestions were—you know—_welcome_. 😄

:-)  My hunch on this one is the scope you've given it is too narrow not that there isn't real problem at the root of it. I'd argue an easy / clear way to make variables fail fast in lots of contexts (not just string interpolation) . A strict label is good for variable which don't need a scope (like environment ones which is where this started).  Variants  of the brace syntax need not be confined to string interpolation...
